sdk/servicebus/service-bus/src/core/messageReceiver.ts

```diff
@@ -848,15 +856,39 @@ export class MessageReceiver extends LinkEntity {
 
   /**
    * Will reconnect the receiver link if necessary.
-   * @param {AmqpError | Error} [receiverError] The receiver error if any.
+   * @param receiverError The receiver error or connection error, if any.
+   * @param connectionDidDisconnect Whether this method is called as a result of a connection disconnect.
    * @returns {Promise<void>} Promise<void>.
    */
-  async onDetached(receiverError?: AmqpError | Error): Promise<void> {
+  async onDetached(receiverError?: AmqpError | Error, causedByDisconnect?: boolean): Promise<void> {
     const connectionId = this._context.namespace.connectionId;
+
+    // User explicitly called `close` on the receiver, so link is already closed
+    // and we can exit early.
+    if (this.wasCloseInitiated) {
+      log.error(
+        `[${connectionId}] Receiver '${this.name}' with address '${this.address}' has had onDetached called after being explictly closed.`
+      );
+      return;
+    }
```

Comment:
The above check on whether user closed the receiver is new. What is the motivation for adding it?
When user calls close, one of the things that happen is that we unset `this._context.streamingReceiver` & remove all listeners on the underlying rhea link. This means
- any further `disconnect` events fired will not call `onDetached`
- no link/session close events fired on receiver link, so no call to `onDetached`

The other possibilities are
- User calling close in between retry attempts here in onDetached
- User calling close when `init()` in onDetached is in progress

Both of these are already handled.

Reply:
It was really to make it easier to write tests for it, since because I'm testing `onDetached` directly in the tests, not all of those assumptions hold true.